### PR TITLE
[MVVM Structure] Part 1 - Add helper modules

### PR DESCRIPTION
**Pull request by manh-t**
https://trello.com/c/AOepchkv

## What happened 🤔
Add some modules to help us easily implement task

## Insight 👀
- Add common-rx and initial domain module
- Fix crash issue when requesting camera permission

### CODE REVIEW

File: data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.nimbl3.data.service.response
+
+import com.google.gson.annotations.SerializedName
+
+data class ErrorResponse(
+    @SerializedName("message")
```
Comment (Tuubz) [inline data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt:6]: Optional: redundant

File: common-rx/src/main/java/co/nimblehq/common/transformers/Transformers.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+import io.reactivex.Observable
+
+object Transformers {
+
+    /**
+     * Emits the latest value of the `source` Observable whenever the `when`
+     * Observable emits.
+     */
+    fun <S, T> takeWhen(`when`: Observable<T>): TakeWhenTransformer<S, T> {
```
Comment (Tuubz) [inline common-rx/src/main/java/co/nimblehq/common/transformers/Transformers.kt:11]: Optional: Should we rename ``when`` to `observable`?

File: domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    private fun isCanceledIOException(error: Throwable) =
+        error is IOException && error.message == MESSAGE_REQUEST_CANCELED
+
+    private fun isCanceledStreamResetException(error: Throwable) =
+        error is StreamResetException && error.message == MESSAGE_STREAM_WAS_RESET_CANCEL
+
+}
+
+private const val MESSAGE_REQUEST_CANCELED = "Canceled"
+private const val MESSAGE_STREAM_WAS_RESET_CANCEL = "stream was reset: CANCEL"
```
Comment (Tuubz) [inline domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt:38]: Optional: Should this be place at the top of the class?

File: common-rx/src/main/java/co/nimblehq/common/Optional.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
+package co.nimblehq.common;
```
[the remark this answers is not in the document]
Comment (luongvo) [reply common-rx/src/main/java/co/nimblehq/common/Optional.java:1]: Why not yeah 💭

File: domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt (hunk 3 above)
Comment (luongvo) [reply domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt:38]: @Tuubz I don't think so, like private methods in a class, we mostly put them on the bottom of the class. Those constants are not the things we would normally like to check when touching into a class for the first time and put them on top would bring annoying tbh 🤔

File: data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt (hunk 1 above)
Comment (luongvo) [reply data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt:6]: @Tuubz there are 2 approach when working with json objects definition:
- only define `SerializedName` for mismatched property names
- define all `SerializedName` for any property names

From my experiences, to avoid mistake, we should take care of defining `SerializedName` for all property names because it would bring a consistent field definitions code format, yeah?
```
data class ErrorResponse(
    @SerializedName("message")
    val message: String,
    @SerializedName("error_code")
    val errorCode: String
)
```
vs
```
data class ErrorResponse(
    val message: String,
    @SerializedName("error_code")
    val errorCode: String
)
```

**Review by luongvo — COMMENTED** [review]
The rest lgtm @manh-t, I will rebase next PR #48

File: domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt (hunk 3 above)
Comment (Tuubz) [reply domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt:38]: Aha, my bad. Seems like I've been in the 7-Eleven project for too long 🙏 😁 I forgot that our guidelines specify to have it at the bottom

File: data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt (hunk 1 above)
Comment (Tuubz) [reply data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt:6]: I love consistency 😍 🆗

Comment (lydiasama) [reply data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt:6]: How about this?
```java;
data class ErrorResponse(
    @Transient
    val message: String,
    @SerializedName("error_code")
    val errorCode: String
)
```

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.jvm/-transient/

Comment (luongvo) [reply data/src/main/java/com/nimbl3/data/service/response/ErrorResponse.kt:6]: @lydiasama we're not discussing how we ignore a field out of a json object (with Transient), instead should we keep defining `@SerializedName` for the same field name and json key name.

**Comment by manh-t** [comment]
Updated 🚀

File: domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt (hunk 3 above)
Comment (lydiasama) [reply domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt:38]: in DeeMoney we put it on top as well 🤦🏻  @Tuubz

Comment (manh-t) [reply domain/src/main/java/co/nimblehq/domain/usecase/base/BaseUseCase.kt:38]: @lydiasama the inspiration is from 7/11 I guess 😆

File: domain/src/main/java/co/nimblehq/domain/usecase/base/SingleUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+                    Single.never()
+                } else {
+                    Single.error(composeError(it))
+                }
+            }
+            .doOnError(::doOnError)
+            .subscribeOn(executionThread)
+            .observeOn(postExecutionThread)
+    }
+
```
Comment (lydiasama) [inline domain/src/main/java/co/nimblehq/domain/usecase/base/SingleUseCase.kt:28]: I think this is a redundant line, I've seen some end of the class has no empty line like this (such as [this line).](https://github.com/nimblehq/android-templates/pull/56/files#diff-1bb3f60c45467134a31713a8de5b38ddf45f8e8519e741ee9be96acf3bfebb10R19) What is our convention then?

Comment (luongvo) [reply domain/src/main/java/co/nimblehq/domain/usecase/base/SingleUseCase.kt:28]: Correct this is a redundant line, applying detekt could resolve this but we could improve this later I think @lydiasama @manh-t

**Review by lydiasama — APPROVED** [review]

File: domain/build.gradle [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+apply from: '../config/jacoco.gradle'
+
+def globalConfig = rootProject.extensions.getByName("ext")
+android {
+    compileSdkVersion globalConfig.getAt("android_compile_sdk_version")
+    defaultConfig {
+        minSdkVersion globalConfig.getAt("android_min_sdk_version")
+        targetSdkVersion globalConfig.getAt("android_target_sdk_version")
+        versionCode globalConfig.getAt("android_version_code")
+        versionName globalConfig.getAt("android_version_name")
```
Comment (Tuubz) [inline domain/build.gradle:14]: Optional: I believe this could be `globalConfig["android_version_name"]`

File: common-rx/src/main/java/co/nimblehq/common/transformers/TakeWhenTransformer.kt [hunk 7]
```diff
@@ -1,11 +1,11 @@
-package com.nimbl3.data.lib.rxjava.transformers
+package co.nimblehq.common.transformers
 
 import io.reactivex.Observable
 import io.reactivex.ObservableSource
 import io.reactivex.ObservableTransformer
 import io.reactivex.functions.BiFunction
 
-class TakeWhenTransformer<S, in T> constructor(private val `when` : Observable<in T>) : ObservableTransformer<S, S> {
+class TakeWhenTransformer<S, in T> constructor(private val `when`: Observable<in T>) : ObservableTransformer<S, S> {
```
Comment (Tuubz) [inline common-rx/src/main/java/co/nimblehq/common/transformers/TakeWhenTransformer.kt:8]: Optional: Shall we rename `when` to `observable` ?

File: domain/build.gradle [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    flavorDimensions "stage"
```
Comment (Tuubz) [inline domain/build.gradle:17]: So, `stage` is referring to `staging`?

File: domain/build.gradle [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    androidExtensions {
+        experimental = true
+    }
+}
+
+dependencies {
+    api project(':data')
+    implementation project(':common-rx')
```
Comment (Tuubz) [inline domain/build.gradle:60]: Hmmm, just wondering. I don't see any references to `project(':domain')` ? Is not it being used, or did I miss something 🙈  ?

File: domain/build.gradle [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
+
+    androidExtensions {
+        experimental = true
+    }
+}
+
+dependencies {
+    api project(':data')
```
Comment (Tuubz) [inline domain/build.gradle:59]: Any reason why this needs to be `api` instead of `implementation`?

File: common-rx/build.gradle [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    // noinspection DifferentStdlibGradleVersion
+    implementation(
+        "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    )
+
+    api(
+        "io.reactivex.rxjava2:rxjava:$rxjava_version",
+        "io.reactivex.rxjava2:rxkotlin:$rxkotlin_version"
+    )
+}
```
Comment (Tuubz) [inline common-rx/build.gradle:14]: Hmmm, does this we can remove the  libraries from `data/build.gradle`? :still_thinking:

Most likely `"io.reactivex.rxjava2:rxandroid:$rxandroid_version"` should be moved in to `common-rx/build.gradle` as well 🙏 

Sorry of this is out of scope 🙈

**Review by luongvo — APPROVED** [review]

File: domain/build.gradle (hunk 8 above)
Comment (luongvo) [reply domain/build.gradle:17]: @Tuubz this is `flavorDimensions`. In multiple modules android project, to make a constraint on build flavor between modules, we need to define and use the same flavorDimensions

File: domain/build.gradle (hunk 9 above)
Comment (luongvo) [reply domain/build.gradle:60]: @Tuubz out of PRs scope https://github.com/nimblehq/android-templates/pull/48/files#diff-51a0b488f963eb0be6c6599bf5df497313877cf5bdff3950807373912ac1cdc9R124-R125

File: domain/build.gradle (hunk 10 above)
Comment (luongvo) [reply domain/build.gradle:59]: @Tuubz to allow app module access to data module components
so instead of 
```
app
    implementation project(':common-rx')
    implementation project(':data')
    implementation project(':domain')

domain
    implementation project(':data')
```
we can define
```
app
    implementation project(':common-rx')
    implementation project(':domain')

domain
    api project(':data')
```

File: domain/build.gradle (hunk 9 above)
Comment (manh-t) [reply domain/build.gradle:60]: yeah, currently we haven't applied use cases in main app yet, it will be coming on other PR 🙏

File: common-rx/build.gradle (hunk 11 above)
Comment (luongvo) [reply common-rx/build.gradle:14]: @Tuubz correct, @manh-t we could update in next PR https://github.com/nimblehq/android-templates/pull/48

Comment (manh-t) [reply common-rx/build.gradle:14]: @luongvo 🆗

File: domain/build.gradle (hunk 8 above)
Comment (Tuubz) [reply domain/build.gradle:17]: Got it, but in the future we should preferably rename it to `staging`? 😋

Comment (luongvo) [reply domain/build.gradle:17]: @Tuubz nooo lol, `stage` here doesn't mean `staging`, actually its name should be `flavor` or something like that
```
flavorDimensions "stage"
    productFlavors {
        production {
            dimension "stage"
        }

        staging {
            dimension "stage"
        }

        development {
            dimension "stage"
        }
    }
```

**Comment by manh-t** [comment]
@Tuubz resolved 🚀

**Review by luongvo — APPROVED** [review]

File: common-rx/build.gradle [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+targetCompatibility = JavaVersion.VERSION_1_8
+sourceCompatibility = JavaVersion.VERSION_1_8
```
Comment (Tuubz) [inline common-rx/build.gradle:17]: Is this doing anything? 🙈 

Android studio tells me "Assignment is not used"

**Review by Tuubz — APPROVED** [review]

File: common-rx/build.gradle (hunk 12 above)
Comment (manh-t) [reply common-rx/build.gradle:17]: @Tuubz my AS doesn't show anything (including build) 😳

Comment (Tuubz) [reply common-rx/build.gradle:17]: lol, I synced the project again. Seems to be fine 🙇
